### ISSUE TITLE
Regression - attribute_will_change! doesn't work for nested resources

### ISSUE THREAD

**Issue by alse** [rails/rails#27956]
### Failing Test

```ruby
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"
  gem "rails", github: "rails/rails"
  gem "arel", github: "rails/arel"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"

ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")

ActiveRecord::Schema.define do
  create_table :users, force: true do |t|
    t.string :name
  end

  create_table :posts, force: true do |t|
    t.integer :user_id
    t.text :body
  end
end

class User < ActiveRecord::Base
  has_many :posts
  accepts_nested_attributes_for :posts
end

class Post < ActiveRecord::Base
  attr_accessor :virtual_input_attribute

  belongs_to :user
  before_save :set_post_body

  def virtual_input_attribute=(value)
    attribute_will_change!(:virtual_input_attribute) unless value.blank?
    @virtual_input_attribute = value
  end

  def set_post_body
    self.body = virtual_input_attribute
  end
end

p ActiveRecord::VERSION::STRING

class BugTest < Minitest::Test
  # Fails on edge Rails. Passes on 5.0.1
  def test_nested_resource_dirty
    user = User.create! name: 'Ruby'
    post = Post.create!(user_id: user.id, body: nil)

    user.update(posts_attributes: { '0' => { id: post.id, virtual_input_attribute: '1' } })

    assert_equal '1', post.reload.body
  end
end
```

### Expected behavior
attribute_will_change! should make nested resources dirty, and the post's body should be updated.

### Actual behavior
attribute_will_change! has no effect for nested resources. The post's body remains nil.

It still works when directly updating the resource. This test passes:

```ruby
  def test_resource_dirty
    user = User.create! name: 'Ruby'
    post = Post.create!(user_id: user.id, body: nil)
    
    post.update(virtual_input_attribute: '1')

    assert_equal '1', post.reload.body
  end
```

### System configuration
**Rails version**:
Edge Rails cabb212976703dac46bacc36addf70a232e96e71
**Ruby version**:
ruby 2.4.0p0

I tracked it down to this commit: https://github.com/rails/rails/commit/16ae3db5a5c6a08383b974ae6c96faac5b4a3c81

**Comment by alse** [comment]
I added a fix above. I think the test will need some work, but let me know if I should open a PR.

**Comment by sgrif** [comment]
If it worked in 5.0.1, we should fix for 5.1. However, I want to deprecate this behavior. The intention is to require the use of `attribute` if you want it to work with the various attribute methods, including dirty.

**Comment by sgrif** [comment]
I'm not fully convinced this was ever intended to work, however.
